File: backend/gemini-proxy/src/main/java/com/tranphuc8a/gemini_proxy/application/ports/output/MessageOutputPort.java
package com.tranphuc8a.gemini_proxy.application.ports.output;

import com.tranphuc8a.gemini_proxy.domain.models.Message;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface MessageOutputPort {

    Message getById(Integer messageId);

    Page<Message> getAll(Pageable pageable);

    Page<Message> getListByConversation(String conversationId, Pageable pageable);

    Message save(Message message);

    void delete(Message message);

    void delete(Integer messageId);

}

File: backend/fastapi/src/application/ports/output/message_output_port.py
from abc import ABC, abstractmethod
from typing import List, Optional

from domain.models.message_domain import MessageDomain


class MessageOutputPort(ABC):
    @abstractmethod
    async def get_list_by_conversation(self, conversation_id: str, after: Optional[str], limit: int, order: str) -> tuple[List[MessageDomain], bool]:
        pass
    
    @abstractmethod
    async def get_all_by_conversation(self, conversation_id: str) -> List[MessageDomain]:
        pass

    @abstractmethod
    async def save(self, message: MessageDomain) -> MessageDomain:
        pass
    
    @abstractmethod
    async def delete(self, message: MessageDomain) -> bool:
        pass
    
    @abstractmethod
    async def delete_by_conversation(self, conversation_id: str) -> bool:
        pass
    
    @abstractmethod
    async def count_by_conversation(self, conversation_id: str) -> int:
        pass
    
    @abstractmethod
    async def get_latest_by_conversation(self, conversation_id: str, count: int) -> List[MessageDomain]:
        pass
    
    @abstractmethod
    async def get_by_id(self, message_id: int) -> MessageDomain:
        pass
    
    @abstractmethod
    async def update(self, message: MessageDomain) -> MessageDomain:
        pass
    
    @abstractmethod
    async def get_by_conversation_and_role(self, conversation_id: str, role: str) -> List[MessageDomain]:
        pass
    